sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeColumnCommand.scala

```diff
@@ -42,17 +42,28 @@ case class AnalyzeColumnCommand(
     if (tableMeta.tableType == CatalogTableType.VIEW) {
       throw new AnalysisException("ANALYZE TABLE is not supported on views.")
     }
-    val sizeInBytes = AnalyzeTableCommand.calculateTotalSize(sessionState, tableMeta)
+    val oldRowCount = tableMeta.stats.flatMap(_.rowCount.map(_.toLong)).getOrElse(-1L)
+    val oldColStats = tableMeta.stats.map(_.colStats).getOrElse(Map.empty)
+    val newSize = AnalyzeTableCommand.calculateTotalSize(sessionState, tableMeta)
 
     // Compute stats for each column
-    val (rowCount, newColStats) = computeColumnStats(sparkSession, tableIdentWithDB, columnNames)
+    val (newRowCount, colStats) = computeColumnStats(sparkSession, tableIdentWithDB, columnNames)
+
+    val newColStats = if (newRowCount == 0) {
+      // No need to keep column stats when there's no data.
+      Map.empty[String, ColumnStat]
+    } else if (newRowCount == oldRowCount) {
```

Comment:
The behaviors are the same for both EXTERNAL and INTERNAL tables. So I just post results for internal table below. It seems hive keeps stale column stats after inserting new data (but totalSize and rowCount in table parameter are updated).
```
hive> create table src (key int);
hive> insert into table src select 1;
hive> desc extended src;
OK
key                 	int                 	                    
	 	 
Detailed Table Information	Table(tableName:src, dbName:default, owner:wzh, createTime:1497380989, lastAccessTime:0, retention:0, sd:StorageDescriptor(cols:[FieldSchema(name:key, type:int, comment:null)], location:hdfs://localhost:9000/user/hive/warehouse/src, inputFormat:org.apache.hadoop.mapred.TextInputFormat, outputFormat:org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat, compressed:false, numBuckets:-1, serdeInfo:SerDeInfo(name:null, serializationLib:org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, parameters:{serialization.format=1}), bucketCols:[], sortCols:[], parameters:{}, skewedInfo:SkewedInfo(skewedColNames:[], skewedColValues:[], skewedColValueLocationMaps:{}), storedAsSubDirectories:false), partitionKeys:[], parameters:{totalSize=2, numRows=1, rawDataSize=1, COLUMN_STATS_ACCURATE={"BASIC_STATS":"true"}, numFiles=1, transient_lastDdlTime=1497381019}, viewOriginalText:null, viewExpandedText:null, tableType:MANAGED_TABLE)

hive> analyze table src compute statistics for columns key;
hive> desc formatted src key;
OK
# col_name            	data_type           	min                 	max                 	num_nulls          distinct_count      	avg_col_len         	max_col_len         	num_trues           	num_falses         comment             
	 	 	 	 	 	 	 	 	 	 
key                 	int                 	1                   	1                   	0                  1                   	                    	                    	                    	                   from deserializer

hive> insert into table src select 2;
hive> desc extended src;
OK
key                 	int                 	                    
	 	 
Detailed Table Information	Table(tableName:src, dbName:default, owner:wzh, createTime:1497380989, lastAccessTime:0, retention:0, sd:StorageDescriptor(cols:[FieldSchema(name:key, type:int, comment:null)], location:hdfs://localhost:9000/user/hive/warehouse/src, inputFormat:org.apache.hadoop.mapred.TextInputFormat, outputFormat:org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat, compressed:false, numBuckets:-1, serdeInfo:SerDeInfo(name:null, serializationLib:org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, parameters:{serialization.format=1}), bucketCols:[], sortCols:[], parameters:{}, skewedInfo:SkewedInfo(skewedColNames:[], skewedColValues:[], skewedColValueLocationMaps:{}), storedAsSubDirectories:false), partitionKeys:[], parameters:{totalSize=4, numRows=2, rawDataSize=2, COLUMN_STATS_ACCURATE={"BASIC_STATS":"true"}, numFiles=2, transient_lastDdlTime=1497381316}, viewOriginalText:null, viewExpandedText:null, tableType:MANAGED_TABLE)	
Time taken: 0.05 seconds, Fetched: 3 row(s)
hive> desc formatted src key;
OK
# col_name            	data_type           	min                 	max                 	num_nulls          distinct_count      	avg_col_len         	max_col_len         	num_trues           	num_falses         comment             
	 	 	 	 	 	 	 	 	 	 
key                 	int                 	1                   	1                   	0                  1                   	                    	                    	                    	                   from deserializer

```